Public/Src/Cache/ContentStore/Distributed/NuCache/RocksDbContentLocationDatabase.cs

```diff
@@ -106,6 +119,55 @@ protected override BoolResult InitializeCore(OperationContext context)
             return result;
         }
 
+        /// <inheritdoc />
+        public override void SetDatabaseMode(bool isDatabaseWriteable)
+        {
+            if (IsGarbageCollectionEnabled != isDatabaseWriteable)
+            {
+                var nextCompactionTimeSpan = isDatabaseWriteable ? _configuration.FullRangeCompactionInterval : Timeout.InfiniteTimeSpan;
+                _compactionTimer?.Change(nextCompactionTimeSpan, Timeout.InfiniteTimeSpan);
+            }
+
+            base.SetDatabaseMode(isDatabaseWriteable);
+        }
+
+        private void FullRangeCompaction(OperationContext context)
+        {
+            lock (ShutdownLock)
+            {
+                if (ShutdownStarted)
+                {
+                    return;
+                }
+            }
+
+            // We can safely ignore the result, as any relevant errors will be logged by the inner operations
+            _keyValueStore.Use(store =>
+            {
+                foreach (var columnFamilyName in new[] { "default", nameof(Columns.ClusterState), nameof(Columns.Metadata) })
+                {
+                    var result = context.PerformOperation(Tracer, () =>
+                    {
+                        store.CompactRange((byte[])null, null, columnFamilyName: columnFamilyName);
+                        return BoolResult.Success;
+                    }, extraStartMessage: $"ColumnFamily={columnFamilyName}");
+
+                    if (!result.Succeeded)
+                    {
+                        break;
+                    }
+                }
+            }).ToBoolResult().IgnoreFailure();
+
+            lock (ShutdownLock)
+            {
+                if (!ShutdownStarted)
+                {
+                    _compactionTimer?.Change(_configuration.FullRangeCompactionInterval, Timeout.InfiniteTimeSpan);

```

Comment:
I don't think you need to use locks here, because ShutdownStarted is backed by cancellation token source that is thread safe.

Reply:
The same is true for both checks: at the beginning and at the end of the method.

---
In reply to: [313644899](https://github.com/microsoft/BuildXL/pull/731#discussion_r313644899) [](ancestors = 313644899)